File: GoVersion/kdtree.go
package main

import (
	"math"
	"math/rand"
)

const (
	MAX_PARTS int     = 7
	THETA     float64 = 0.3
)

type KDTree struct {
	// For leaves
	num_parts int
	particles [MAX_PARTS]int

	// For internal nodes
	split_dim int
	split_val float64
	m         float64
	cm        [3]float64
	size      float64
	left      int
	right     int
}

func empty_leaf() KDTree {
	return KDTree{
		0,
		[MAX_PARTS]int{},
		-1,
		0.0,
		0.0,
		[3]float64{0.0, 0.0, 0.0},
		0.0,
		-1,
		-1,
	}
}

func allocate_node_vec(num_parts int) []KDTree {
	num_nodes := 2 * (num_parts/(MAX_PARTS-1) + 1)
	ret := make([]KDTree, num_nodes)
	return ret
}

// Returns the index of the last Node used in the construction.
func build_tree(
	indices []int,
	start int,
	end int,
	particles []Particle,
	cur_node int,
	nodes []KDTree,
) int {
	// println!("start = {} end = {} cur_node = {}", start, end, cur_node);
	np := end - start
	// println!("s = {}, e = {}, cn = {}", start, end, cur_node);
	if np <= MAX_PARTS {
		for cur_node >= len(nodes) {
			nodes = append(nodes, empty_leaf())
		}
		nodes[cur_node].num_parts = np
		for i := 0; i < np; i++ {
			nodes[cur_node].particles[i] = indices[start+i]
		}
		return cur_node
	} else {
		// Pick split dim and value
		min := [3]float64{1e100, 1e100, 1e100}
		max := [3]float64{-1e100, -1e100, -1e100}
		m := 0.0
		cm := [3]float64{0.0, 0.0, 0.0}
		for i := start; i < end; i++ {
			m += particles[indices[i]].m
			cm[0] += particles[indices[i]].m * particles[indices[i]].p[0]
			cm[1] += particles[indices[i]].m * particles[indices[i]].p[1]
			cm[2] += particles[indices[i]].m * particles[indices[i]].p[2]
			min[0] = math.Min(min[0], particles[indices[i]].p[0])
			min[1] = math.Min(min[1], particles[indices[i]].p[1])
			min[2] = math.Min(min[2], particles[indices[i]].p[2])
			max[0] = math.Max(max[0], particles[indices[i]].p[0])
			max[1] = math.Max(max[1], particles[indices[i]].p[1])
			max[2] = math.Max(max[2], particles[indices[i]].p[2])
		}
		cm[0] /= m
		cm[1] /= m
		cm[2] /= m
		split_dim := 0
		for dim := 1; dim < 3; dim++ {
			if max[dim]-min[dim] > max[split_dim]-min[split_dim] {
				split_dim = dim
			}
		}
		size := max[split_dim] - min[split_dim]

		// Partition particles on split_dim
		mid := (start + end) / 2
		s := start
		e := end
		for s+1 < e {
			pivot := s + rand.Int()%(e-s)
			tmp := indices[s]
			indices[s] = indices[pivot]
			indices[pivot] = tmp
			low := s + 1
			high := e - 1
			for low <= high {
				if particles[indices[low]].p[split_dim] < particles[indices[s]].p[split_dim] {
					low += 1
				} else {
					tmp = indices[low]
					indices[low] = indices[high]
					indices[high] = tmp
					high -= 1
				}
			}
			tmp = indices[s]
			indices[s] = indices[high]
			indices[high] = tmp
			if high < mid {
				s = high + 1
			} else if high > mid {
				e = high
			} else {
				s = e
			}
		}
		split_val := particles[indices[mid]].p[split_dim]

		// Recurse on children and build this node.
		left := build_tree(indices, start, mid, particles, cur_node+1, nodes)
		right := build_tree(indices, mid, end, particles, left+1, nodes)

		for cur_node >= len(nodes) {
			nodes = append(nodes, empty_leaf())
		}
		nodes[cur_node].num_parts = 0
		nodes[cur_node].split_dim = split_dim
		nodes[cur_node].split_val = split_val
		nodes[cur_node].m = m
		nodes[cur_node].cm = cm
		nodes[cur_node].size = size
		nodes[cur_node].left = cur_node + 1
		nodes[cur_node].right = left + 1

		return right
	}
}

func accel_recur(cur_node int, p int, particles []Particle, nodes []KDTree) [3]float64 {
	// println!("accel {}", cur_node);
	if nodes[cur_node].num_parts > 0 {
		acc := [3]float64{0.0, 0.0, 0.0}
		for i := 0; i < nodes[cur_node].num_parts; i++ {
			if nodes[cur_node].particles[i] != p {
				pp_acc := Calc_pp_accel(&particles[p], &particles[nodes[cur_node].particles[i]])
				acc[0] += pp_acc[0]
				acc[1] += pp_acc[1]
				acc[2] += pp_acc[2]
			}
		}
		return acc
	} else {
		dx := particles[p].p[0] - nodes[cur_node].cm[0]
		dy := particles[p].p[1] - nodes[cur_node].cm[1]
		dz := particles[p].p[2] - nodes[cur_node].cm[2]
		dist_sqr := dx*dx + dy*dy + dz*dz
		// println!("dist = {}, size = {}", dist, nodes[cur_node].size);
		if nodes[cur_node].size*nodes[cur_node].size < THETA*THETA*dist_sqr {
			dist := math.Sqrt(dist_sqr)
			magi := -nodes[cur_node].m / (dist_sqr * dist)
			return [3]float64{dx * magi, dy * magi, dz * magi}
		} else {
			left_acc := accel_recur(nodes[cur_node].left, p, particles, nodes)
			right_acc := accel_recur(nodes[cur_node].right, p, particles, nodes)
			return [3]float64{left_acc[0] + right_acc[0], left_acc[1] + right_acc[1], left_acc[2] + right_acc[2]}
		}
	}
}

func calc_accel(p int, particles []Particle, nodes []KDTree) [3]float64 {
	return accel_recur(0, p, particles, nodes)
}

func Simple_sim(bodies []Particle, dt float64, steps int) {
	acc := make([][3]float64, len(bodies))
	tree := allocate_node_vec(len(bodies))
	indices := make([]int, len(bodies))

	for step := 0; step < steps; step++ {
		// if step % 100 == 0 {
		//     let elapsed_secs = time.elapsed().as_nanos() as f64 / 1e9;
		//     println!("Step = {}, duration = {}, n = {}, nodes = {}", step, elapsed_secs, bodies.len(), tree.len());
		//     time = Instant::now();
		// }
		for i := 0; i < len(bodies); i++ {
			indices[i] = i
		}
		build_tree(indices, 0, len(bodies), bodies, 0, tree)
		// if step % 10 == 0 {
		//     print_tree(step, &tree, &bodies);
		// }
		for i := 0; i < len(bodies); i++ {
			acc[i] = calc_accel(i, bodies, tree)
		}
		for i := 0; i < len(bodies); i++ {
			bodies[i].v[0] += dt * acc[i][0]
			bodies[i].v[1] += dt * acc[i][1]
			bodies[i].v[2] += dt * acc[i][2]
			dx := dt * bodies[i].v[0]
			dy := dt * bodies[i].v[1]
			dz := dt * bodies[i].v[2]
			bodies[i].p[0] += dx
			bodies[i].p[1] += dy
			bodies[i].p[2] += dz
			acc[i][0] = 0.0
			acc[i][1] = 0.0
			acc[i][2] = 0.0
		}
	}
}

// func print_tree(step int, tree []KDTree, particles []Particle) {
//     let mut file = File::create(format!("tree{}.txt", step))?;

//     file.write_fmt(format_args!("{}\n", tree.len()))?;
//     for n in tree {
//         if n.num_parts > 0 {
//             file.write_fmt(format_args!("L {}\n", n.num_parts))?;
//             for i in 0..n.num_parts {
//                 let p = n.particles[i];
//                 file.write_fmt(format_args!(
//                     "{} {} {}\n",
//                     particles[p].p[0], particles[p].p[1], particles[p].p[2]
//                 ))?;
//             }
//         } else {
//             file.write_fmt(format_args!(
//                 "I {} {} {} {}\n",
//                 n.split_dim, n.split_val, n.left, n.right
//             ))?;
//         }
//     }

//     Ok(())
// }

// fn recur_test_tree_struct(
//     node: usize,
//     nodes: &Vec<KDTree>,
//     particles: &Vec<Particle>,
//     mut min: [f64; 3],
//     mut max: [f64; 3],
// ) {
//     if nodes[node].num_parts > 0 {
//         for index in 0..nodes[node].num_parts {
//             let i = nodes[node].particles[index];
//             for dim in 0..2 {
//                 assert!(
//                     particles[i].p[dim] >= min[dim],
//                     "Particle dim {} is below min. i={} p={} min={}",
//                     dim,
//                     i,
//                     particles[i].p[dim],
//                     min[dim]
//                 );
//                 assert!(
//                     particles[i].p[dim] < max[dim],
//                     "Particle dim {} is above max. i={} p={} max={}",
//                     dim,
//                     i,
//                     particles[i].p[dim],
//                     max[dim]
//                 );
//             }
//         }
//     } else {
//         let split_dim = nodes[node].split_dim;
//         let tmin = min[split_dim];
//         let tmax = max[split_dim];
//         max[split_dim] = nodes[node].split_val;
//         recur_test_tree_struct(nodes[node].left, nodes, particles, min, max);
//         max[split_dim] = tmax;
//         min[split_dim] = nodes[node].split_val;
//         recur_test_tree_struct(nodes[node].right, nodes, particles, min, max);
//         min[split_dim] = tmin;
//     }
// }

// #[cfg(test)]
// mod tests {
//     use crate::{array_kd_tree, array_particle};

//     #[test]
//     fn single_node() {
//         let parts = array_particle::two_bodies();
//         let mut node_vec = array_kd_tree::allocate_node_vec(parts.len());
//         assert_eq!(node_vec.len(), 2);
//         let mut indices: Vec<usize> = (0..parts.len()).collect();
//         array_kd_tree::build_tree(&mut indices, 0, parts.len(), &parts, 0, &mut node_vec);
//         assert_eq!(node_vec[0].num_parts, parts.len());
//     }

//     #[test]
//     fn two_leaves() {
//         let parts = array_particle::circular_orbits(11);
//         let mut node_vec = array_kd_tree::allocate_node_vec(parts.len());
//         assert_eq!(node_vec.len(), 6);
//         let mut indices: Vec<usize> = (0..parts.len()).collect();
//         array_kd_tree::build_tree(&mut indices, 0, parts.len(), &parts, 0, &mut node_vec);
//         array_kd_tree::recur_test_tree_struct(
//             0,
//             &node_vec,
//             &parts,
//             [-1e100, -1e100, -1e100],
//             [1e100, 1e100, 1e100],
//         );
//         assert_eq!(node_vec[0].num_parts, 0);
//         assert_eq!(node_vec[1].num_parts + node_vec[2].num_parts, 12);
//     }

//     #[test]
//     fn big_solar() {
//         let parts = array_particle::circular_orbits(5000);
//         let mut node_vec = array_kd_tree::allocate_node_vec(parts.len());
//         let mut indices: Vec<usize> = (0..parts.len()).collect();
//         array_kd_tree::build_tree(&mut indices, 0, parts.len(), &parts, 0, &mut node_vec);
//         array_kd_tree::recur_test_tree_struct(
//             0,
//             &node_vec,
//             &parts,
//             [-1e100, -1e100, -1e100],
//             [1e100, 1e100, 1e100],
//         );
//     }

//     #[test]
//     fn big_solar_with_steps() {
//         let mut parts = array_particle::circular_orbits(5000);
//         array_kd_tree::simple_sim(&mut parts, 1e-3, 10);

//         let mut node_vec = array_kd_tree::allocate_node_vec(parts.len());
//         let mut indices: Vec<usize> = (0..parts.len()).collect();
//         array_kd_tree::build_tree(&mut indices, 0, parts.len(), &parts, 0, &mut node_vec);
//         array_kd_tree::recur_test_tree_struct(
//             0,
//             &node_vec,
//             &parts,
//             [-1e100, -1e100, -1e100],
//             [1e100, 1e100, 1e100],
//         );
//     }
// }

File: Sequential/RustVersion/src/kd_tree.rs
use std::{fs::File, io::Write};

use core_simd::simd::*;

use crate::simd_particle::*;

use super::simd_particle::Particle;

const MAX_PARTS: usize = 7;
const THETA: f64 = 0.3;
const NEGS: [usize; MAX_PARTS] = [usize::MAX; MAX_PARTS];

#[derive(Clone, Copy)]
pub enum KDTree {
    Leaf {
        num_parts: usize,
        leaf_parts: [usize; MAX_PARTS]
    },

    Internal {
        split_dim: usize,
        split_val: f64,
        m: f64,
        cm: f64x4,
        size: f64,
        left: usize,
        right: usize
    }
}

impl KDTree {
    pub fn leaf<'a>(num_parts: usize, particles: [usize; MAX_PARTS]) -> KDTree {
        KDTree::Leaf {
            num_parts: num_parts,
            leaf_parts: particles,
        }
    }
}

pub fn allocate_node_vec(num_parts: usize) -> Vec<KDTree> {
    let num_nodes = 2 * (num_parts / (MAX_PARTS - 1) + 1);
    let mut ret = Vec::new();
    ret.resize(num_nodes, KDTree::leaf(0, NEGS));
    ret
}

// Returns the index of the last Node used in the construction.
pub fn build_tree(
    indices: &mut Vec<usize>,
    start: usize,
    end: usize,
    particles: &Vec<Particle>,
    cur_node: usize,
    nodes: &mut Vec<KDTree>,
) -> usize {
    // println!("start = {} end = {} cur_node = {}", start, end, cur_node);
    let np = end - start;
    // println!("s = {}, e = {}, cn = {}", start, end, cur_node);
    if np <= MAX_PARTS {
        if cur_node >= nodes.len() {
            nodes.resize(cur_node + 1, KDTree::leaf(0, NEGS));
        }
        let mut parts = [0; MAX_PARTS];
        for i in 0..np {
            parts[i] = indices[start + i]
        }
        nodes[cur_node] = KDTree::Leaf { num_parts: np, leaf_parts: parts };
        cur_node
    } else {
        // Pick split dim and value
        let mut min = f64x4::splat(1e100);
        let mut max = f64x4::splat(-1e100);
        let mut m = 0.0;
        let mut cm = f64x4::splat(0.0);
        for i in start..end {
            m += particles[indices[i]].m;
            cm += f64x4::splat(particles[indices[i]].m) * particles[indices[i]].p;
            min = f64x4::simd_min(min, particles[indices[i]].p);
            max = f64x4::simd_max(max, particles[indices[i]].p);
        }
        cm /= f64x4::splat(m);
        let mut split_dim = 0;
        for dim in 1..3 {
            if max[dim] - min[dim] > max[split_dim] - min[split_dim] {
                split_dim = dim
            }
        }
        let size = max[split_dim] - min[split_dim];

        // Partition particles on split_dim
        let mid = (start + end) / 2;
        let mut s = start;
        let mut e = end;
        while s + 1 < e {
            let pivot = fastrand::usize(s..e);
            indices.swap(s, pivot);
            let mut low = s + 1;
            let mut high = e - 1;
            while low <= high {
                if particles[indices[low]].p[split_dim] < particles[indices[s]].p[split_dim] {
                    low += 1;
                } else {
                    indices.swap(low, high);
                    high -= 1;
                }
            }
            indices.swap(s, high);
            use std::cmp::Ordering::*;
            match high.cmp(&mid) {
                Less => s = high + 1,
                Greater => e = high,
                Equal => s = e,
            }
        }
        let split_val = particles[indices[mid]].p[split_dim];

        // Recurse on children and build this node.
        let left = build_tree(indices, start, mid, particles, cur_node + 1, nodes);
        let right = build_tree(indices, mid, end, particles, left + 1, nodes);

        if cur_node >= nodes.len() {
            nodes.resize(cur_node + 1, KDTree::leaf(0, NEGS));
        }
        nodes[cur_node] = KDTree::Internal { split_dim, split_val, m, cm, size, left: cur_node + 1, right: left+1 };

        right
    }
}

fn accel_recur(cur_node: usize, p: usize, particles: &Vec<Particle>, nodes: &Vec<KDTree>) -> f64x4 {
    // println!("accel {}", cur_node);
    match nodes[cur_node] {
        KDTree::Leaf { num_parts, leaf_parts} => {
            let mut acc = f64x4::splat(0.0);
            for i in 0..num_parts {
                if leaf_parts[i] != p {
                    acc += calc_pp_accel(&particles[p], &particles[leaf_parts[i]]);
                }
            }
            acc
        }
        KDTree::Internal { m, cm, size, left, right, .. } => {
            let dp = particles[p].p - cm;
            let dp2 = dp * dp;
            let dist_sqr = dp2.reduce_sum();
            // println!("dist = {}, size = {}", dist, size);
            if size * size < THETA * THETA * dist_sqr {
                let dist = f64::sqrt(dist_sqr);
                let magi = f64x4::splat(-m / (dist_sqr * dist));
                dp * magi
            } else {
                accel_recur(left, p, particles, nodes)
                    + accel_recur(right, p, particles, nodes)
            }
        }
    }
}

pub fn calc_accel(p: usize, particles: &Vec<Particle>, nodes: &Vec<KDTree>) -> f64x4 {
    accel_recur(0, p, particles, nodes)
}

pub fn simple_sim(bodies: &mut Vec<Particle>, dt: f64, steps: i64) {
    let dt_vec = f64x4::splat(dt);
    let mut acc = Vec::new();
    for _ in 0..bodies.len() {
        acc.push(f64x4::splat(0.0))
    }
    // let mut time = Instant::now();
    let mut tree = allocate_node_vec(bodies.len());
    let mut indices: Vec<usize> = (0..bodies.len()).collect();

    for _ in 0..steps {
        // if step % 100 == 0 {
        //     let elapsed_secs = time.elapsed().as_nanos() as f64 / 1e9;
        //     println!("Step = {}, duration = {}, n = {}, nodes = {}", step, elapsed_secs, bodies.len(), tree.len());
        //     time = Instant::now();
        // }
        for (new_ix, old_ix) in indices.iter_mut().enumerate().take(bodies.len()) {
            *old_ix = new_ix;
        }

        build_tree(&mut indices, 0, bodies.len(), bodies, 0, &mut tree);
        // if step % 10 == 0 {
        //     print_tree(step, &tree, &bodies);
        // }
        for (index, item) in acc.iter_mut().enumerate().take(bodies.len()) {
            *item = calc_accel(index, bodies, &tree)
        }

        for i in 0..bodies.len() {
            bodies[i].v += dt_vec * acc[i];
            let dp = dt_vec * bodies[i].v;
            bodies[i].p += dp;
            acc[i] = f64x4::splat(0.0);
        }
    }
}

#[allow(dead_code)]
fn print_tree(step: i64, tree: &Vec<KDTree>, particles: &[Particle]) -> std::io::Result<()> {
    let mut file = File::create(format!("tree{step}.txt"))?;

    file.write_fmt(format_args!("{}\n", tree.len()))?;
    for n in tree {
        match n {
            KDTree::Leaf { num_parts, leaf_parts } => {
                file.write_fmt(format_args!("L {}\n", num_parts))?;
                for i in 0..*num_parts {
                    let p = leaf_parts[i];
                    file.write_fmt(format_args!(
                        "{} {} {}\n",
                        particles[p].p[0], particles[p].p[1], particles[p].p[2]
                    ))?;
                }
            }
            KDTree::Internal { split_dim, split_val, left, right, .. } => {
                file.write_fmt(format_args!(
                    "I {} {} {} {}\n",
                    split_dim, split_val, left, right
                ))?;
            }
        }
    }

    Ok(())
}

#[allow(dead_code)]
fn recur_test_tree_struct(
    node: usize,
    nodes: &Vec<KDTree>,
    particles: &Vec<Particle>,
    mut min: f64x4,
    mut max: f64x4,
) {
    match nodes[node] {
        KDTree::Leaf { num_parts, leaf_parts } => {
            for index in 0..num_parts {
                let i = leaf_parts[index];
                for dim in 0..2 {
                    assert!(
                        particles[i].p[dim] >= min[dim],
                        "Particle dim {} is below min. i={} p={} min={}",
                        dim,
                        i,
                        particles[i].p[dim],
                        min[dim]
                    );
                    assert!(
                        particles[i].p[dim] < max[dim],
                        "Particle dim {} is above max. i={} p={} max={}",
                        dim,
                        i,
                        particles[i].p[dim],
                        max[dim]
                    );
                }
            }
        }
        KDTree::Internal { split_dim, split_val, left, right, .. } => {
            let split_dim = split_dim;
            let tmin = min[split_dim];
            let tmax = max[split_dim];
            max[split_dim] = split_val;
            recur_test_tree_struct(left, nodes, particles, min, max);
            max[split_dim] = tmax;
            min[split_dim] = split_val;
            recur_test_tree_struct(right, nodes, particles, min, max);
            min[split_dim] = tmin;
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::{kd_tree, simd_particle};
    use core_simd::simd::*;

    #[test]
    fn single_node() {
        let parts = simd_particle::two_bodies();
        let mut node_vec = kd_tree::allocate_node_vec(parts.len());
        assert_eq!(node_vec.len(), 2);
        let mut indices: Vec<usize> = (0..parts.len()).collect();
        kd_tree::build_tree(&mut indices, 0, parts.len(), &parts, 0, &mut node_vec);
        match node_vec[0] {
            kd_tree::KDTree::Leaf { num_parts, .. } => assert_eq!(num_parts, parts.len()),
            _ => assert!(false, "Root isn't leaf of right size when small.")
        };
    }

    #[test]
    fn two_leaves() {
        let parts = simd_particle::circular_orbits(11);
        let mut node_vec = kd_tree::allocate_node_vec(parts.len());
        assert_eq!(node_vec.len(), 6);
        let mut indices: Vec<usize> = (0..parts.len()).collect();
        kd_tree::build_tree(&mut indices, 0, parts.len(), &parts, 0, &mut node_vec);
        kd_tree::recur_test_tree_struct(
            0,
            &node_vec,
            &parts,
            f64x4::splat(-1e100),
            f64x4::splat(1e100),
        );
        assert!(std::matches!(node_vec[0], kd_tree::KDTree::Internal { .. }));
        match (node_vec[1], node_vec[2]) {
            (kd_tree::KDTree::Leaf { num_parts: n1, ..}, kd_tree::KDTree::Leaf {num_parts: n2, ..}) => {
                assert_eq!(n1 + n2, 12);
            }
            _ => assert!(false, "Node vectors weren't leaves.")
        }
    }

    #[test]
    fn big_solar() {
        let parts = simd_particle::circular_orbits(5000);
        let mut node_vec = kd_tree::allocate_node_vec(parts.len());
        let mut indices: Vec<usize> = (0..parts.len()).collect();
        kd_tree::build_tree(&mut indices, 0, parts.len(), &parts, 0, &mut node_vec);
        kd_tree::recur_test_tree_struct(
            0,
            &node_vec,
            &parts,
            f64x4::splat(-1e100),
            f64x4::splat(1e100),
        );
    }

    #[test]
    fn big_solar_with_steps() {
        let mut parts = simd_particle::circular_orbits(5000);
        kd_tree::simple_sim(&mut parts, 1e-3, 10);

        let mut node_vec = kd_tree::allocate_node_vec(parts.len());
        let mut indices: Vec<usize> = (0..parts.len()).collect();
        kd_tree::build_tree(&mut indices, 0, parts.len(), &parts, 0, &mut node_vec);
        kd_tree::recur_test_tree_struct(
            0,
            &node_vec,
            &parts,
            f64x4::splat(-1e100),
            f64x4::splat(1e100),
        );
    }
}
